File: BEmu/HistoricalDataRequest/ElementHistoricFieldDataArray.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="HistoricalDataRequest/ElementHistoricFieldDataArray.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.HistoricalDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementHistoricFieldDataArray : Element
    {
        private readonly List<ElementHistoricFieldData> _fieldData;

        internal ElementHistoricFieldDataArray(Dictionary<DateTime, Dictionary<string, object>> fieldData)
        {
            this._fieldData = new List<ElementHistoricFieldData>();

            foreach (var dtDate in fieldData)
	        {
                ElementHistoricFieldData elmFieldData = new ElementHistoricFieldData(dtDate.Key, dtDate.Value);
                this._fieldData.Add(elmFieldData);
	        }
        }

        public override Element GetValueAsElement(int index) { return this._fieldData[index]; }
        public override object this[int index] { get { return this._fieldData[index]; } }

        public override Name Name { get { return new Name("fieldData"); } }
        public override int NumElements { get { return 0; } }
        public override int NumValues { get { return this._fieldData.Count; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsArray { get { return true; } }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}{1}[] = {{{2}", tabs, this.Name, Environment.NewLine);
            foreach (var item in this._fieldData)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);

            return result;
        }
    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/HistoricalDataRequest/ElementHistoricFieldDataArray.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.HistoricalDataRequest.ElementHistoricFieldDataArray.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu.HistoricalDataRequest;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import com.bemu.BEmu.Datetime;
import com.bemu.BEmu.Name;
import com.bemu.BEmu.Element;

public class ElementHistoricFieldDataArray extends ElementParent
{
	private final List<ElementHistoricFieldData> _fieldData;
	
	ElementHistoricFieldDataArray(Map<Datetime, Map<String, Object>> fieldData)
	{
		this._fieldData = new ArrayList<ElementHistoricFieldData>();
		
		for (Map.Entry<Datetime, Map<String, Object>> item : fieldData.entrySet())
		{
			ElementHistoricFieldData elmFieldData = new ElementHistoricFieldData(item.getKey(), item.getValue());
			this._fieldData.add(elmFieldData);
		}
	}
	
	public Element getValueAsElement(int index)
	{
		return this._fieldData.get(index);
	}
	
    public Name name()
    {
    	return new Name("fieldData");
    }
    
    public int numValues()
    {
    	return this._fieldData.size();
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return true;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    protected StringBuilder prettyPrint(int tabIndent)
    {
        String tabs = com.bemu.BEmu.types.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%s%s[] = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        
        for(int i = 0; i < this._fieldData.size(); i++)
        {
        	result.append(this._fieldData.get(i).prettyPrint(tabIndent + 1));
        }

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
    }
}
